src/main/java/com/flagsmith/mappers/EngineMappers.java

```
@@ -0,0 +1,463 @@
+import com.flagsmith.flagengine.EvaluationContext;
+import com.flagsmith.flagengine.FeatureContext;
+import com.flagsmith.flagengine.FeatureValue;
+import com.flagsmith.flagengine.FlagResult;
+import com.flagsmith.flagengine.IdentityContext;
+import com.flagsmith.flagengine.SegmentCondition;
+import com.flagsmith.flagengine.SegmentContext;
+import com.flagsmith.flagengine.SegmentRule;
+import com.flagsmith.flagengine.Segments;
+import com.flagsmith.flagengine.Traits;
+import com.flagsmith.flagengine.segments.constants.SegmentConditions;
+import com.flagsmith.models.FeatureMetadata;
+import com.flagsmith.models.Flag;
+import com.flagsmith.models.SegmentMetadata;
+import com.flagsmith.models.environments.EnvironmentModel;
+import com.flagsmith.models.features.FeatureModel;
+import com.flagsmith.models.features.FeatureSegmentModel;
+import com.flagsmith.models.features.FeatureStateModel;
+import com.flagsmith.models.features.MultivariateFeatureStateValueModel;
+import com.flagsmith.models.identities.IdentityModel;
+import com.flagsmith.models.projects.ProjectModel;
+import com.flagsmith.models.segments.SegmentConditionModel;
+import com.flagsmith.models.segments.SegmentModel;
+import com.flagsmith.models.segments.SegmentRuleModel;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * EngineMappers
+ *
+ * <p>Utility class for mapping JSON data to flag engine context objects.
+ */
+public class EngineMappers {
+  /**
+   * Maps FlagResult to Flag.
+   * Returns null if metadata is missing or invalid.
+   *
+   * @param flagResult the flag result
+   * @return the mapped flag or null
+   */
+  public static Flag mapFlagResultToFlag(
+      FlagResult flagResult
+  ) {
+    FeatureMetadata metadata;
+
+    metadata = MapperFactory.getMapper()
+        .convertValue(flagResult.getMetadata(), FeatureMetadata.class);
+    
+    if (metadata == null || metadata.getFlagsmithId() == null) {
```

Do we expect metadata not to exist in a `FlagResult`?

---

I guess we'd rather deal with a reported NPE than silently skip flags... Corrected in 183cdf9e102a448d1da2b4bc7ccb2f032078024d.

---

...fixed the test in 7d1c948d24ae9239bcc267c519b91fd9da24ac40.

---

This feels like essentially the same conversation that happened [here](https://github.com/Flagsmith/flagsmith-python-client/pull/163/files/9b23e749ec504eaa8777401aa8368cd495c966b2#r2437217752). In which case, I think raising a NPE is fine since it is a scenario that we don't expect to happen.

---

I tend to agree @matthewelwell — feel free to resolve.